### PR TITLE
chore:  Upgrade ESLint from v7 to v8

### PR DESCRIPTION
**Pull request by nzdjb**
From #3623

Work remaining:
- [x] upgrade eslint-plugin-react (once released)
- [x] ~~upgrade eslint-plugin-prettier (once released)~~
- [x] ~~upgrade eslint-plugin-jest-dom (once released)~~
- [x] upgrade eslint-plugin-jsx-a11y (once released)
- [x] resolve any linting issues surfaced by new versions

### CODE REVIEW

**Comment by nzdjb** [comment]
I initially thought eslint-plugin-prettier and eslint-plugin-jest-dom needed updates, but after the updates to the others, the issues from these seem to have been resolved.

**Comment by nzdjb** [comment]
New issues that appear under ESLint 8 that weren't there with ESLint 7:

```
/workspaces/redwood/packages/forms/src/index.tsx
  479:7  warning  'name' is missing in props validation            react/prop-types
  480:7  warning  'id' is missing in props validation              react/prop-types
  482:7  warning  'errorClassName' is missing in props validation  react/prop-types
  483:7  warning  'errorStyle' is missing in props validation      react/prop-types
  487:7  warning  'validation' is missing in props validation      react/prop-types
  488:7  warning  'onBlur' is missing in props validation          react/prop-types
  489:7  warning  'onChange' is missing in props validation        react/prop-types
  528:7  warning  'name' is missing in props validation            react/prop-types
  529:7  warning  'id' is missing in props validation              react/prop-types
  531:7  warning  'errorClassName' is missing in props validation  react/prop-types
  532:7  warning  'errorStyle' is missing in props validation      react/prop-types
  536:7  warning  'validation' is missing in props validation      react/prop-types
  537:7  warning  'onBlur' is missing in props validation          react/prop-types
  538:7  warning  'onChange' is missing in props validation        react/prop-types
  577:7  warning  'name' is missing in props validation            react/prop-types
  578:7  warning  'id' is missing in props validation              react/prop-types
  580:7  warning  'errorClassName' is missing in props validation  react/prop-types
  581:7  warning  'errorStyle' is missing in props validation      react/prop-types
  585:7  warning  'validation' is missing in props validation      react/prop-types
  586:7  warning  'onBlur' is missing in props validation          react/prop-types
  587:7  warning  'onChange' is missing in props validation        react/prop-types
  696:7  warning  'name' is missing in props validation            react/prop-types
  697:7  warning  'id' is missing in props validation              react/prop-types
  699:7  warning  'errorClassName' is missing in props validation  react/prop-types
  700:7  warning  'errorStyle' is missing in props validation      react/prop-types
  704:7  warning  'validation' is missing in props validation      react/prop-types
  705:7  warning  'onBlur' is missing in props validation          react/prop-types
  706:7  warning  'onChange' is missing in props validation        react/prop-types

/workspaces/redwood/packages/internal/src/generate/graphqlSchema.ts
  48:25  error  Unsafe usage of optional chaining. If it short-circuits with 'undefined' the evaluation will throw TypeError  no-unsafe-optional-chaining

/workspaces/redwood/packages/router/src/links.tsx
   64:10  warning  'onClick' is missing in props validation  react/prop-types
  105:58  warning  'onClick' is missing in props validation  react/prop-types

✖ 31 problems (1 error, 30 warnings)
```

**Comment by nzdjb** [comment]
Have now resolved all the new linting issues.

The changes in packages/forms/src/index.tsx could do with a careful look from someone familiar with the intention of the typing. It wasn't entirely clear what the reason was for 'type' was being omitted and that 'id' not being defined. However, linting and tests both pass.

The other changes should be straightforward.

**Comment by dac09** [comment]
Requesting a HODL ✋ on this till we merge in the babel changes that are going in #3719 please!

**Comment by thedavidprice** [comment]
@nzdjb Fantastic work! I'll queue this up for review to include in the next release. Thank you.

Roger that @dac09 👍

File: packages/internal/src/generate/graphqlSchema.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ export const generateGraphQLSchema = async () => {
     // This tries to clean up the output of those errors.
     console.error(e)
     console.error(chalk.red('Error parsing SDLs or Schema'))
-    for (const error of e?.errors) {
+    for (const error of e.errors ?? []) {
```
Comment (jtoar) [inline packages/internal/src/generate/graphqlSchema.ts:48]: Nice!

**Review by jtoar — APPROVED** [review]
Hey @nzdjb thanks for not only tracking the changes it'd take for us to upgrade to ESLint v8, but then also upgrading us as well! 🚀 

I'm good with merging this as is, just had one question about TS in forms. The TS I did in the forms package is definitely not great but I'm seeing one type hint (for `type`) now that I feel like shouldn't be there:

Before:

image

After:

image

I'd just like to know if you thought it was correct show type as a prop when we do nothing with it in `TextAreaField`, etc?

**Comment by thedavidprice** [comment]
Great work on this! I'm merging now. If there's need to make an adjustment based on @jtoar question above, we can handle in a follow-up PR.

Thanks, everyone. 🚀

**Comment by nzdjb** [comment]
> I'd just like to know if you thought it was correct show type as a prop when we do nothing with it in TextAreaField, etc?

Probably not correct, no, but was the minimal change to resolve the issue. I'll take another bash at it and see what can be done about omitting it.

File: packages/router/src/links.tsx [hunk 2]
```diff
@@ -56,6 +56,7 @@ const useMatch = (pathname: string, options?: UseMatchOptions) => {
 
 interface LinkProps {
   to: string
+  onClick: React.MouseEventHandler<HTMLAnchorElement>
```
Comment (callingmedic911) [inline packages/router/src/links.tsx:59]: I think we should set onClick as an optional prop. `onClick?:..`

File: packages/router/src/links.tsx [hunk 3]
```diff
@@ -95,6 +96,7 @@ interface NavLinkProps {
   to: string
   activeClassName: string
   activeMatchParams?: FlattenSearchParams
+  onClick: React.MouseEventHandler<HTMLAnchorElement>
```
Comment (callingmedic911) [inline packages/router/src/links.tsx:99]: Same here, this could be optional.

**Review by callingmedic911 — COMMENTED** [review]
@nzdjb Thanks for all the work here! 🔥 While I was working on router stuff I came across a TS error.
Screenshot 2021-11-22 at 12 45 42 AM

I've added my suggestion as a fix, have a look whenever you get the chance. Feel free to include it in your ongoing forms related PR.

**Comment by nzdjb** [comment]
> @nzdjb Thanks for all the work here! 🔥 While I was working on router stuff I came across a TS error. Screenshot 2021-11-22 at 12 45 42 AM
> 
> I've added my suggestion as a fix, have a look whenever you get the chance. Feel free to include it in your ongoing forms related PR.

Whoops! Sorry about that! 😞

Have added your suggested changes to #3762.

**Comment by thedavidprice** [comment]
@nzdjb we have a monthly meetup for Redwood Contributors, and it would be a pleasure to add you to the group invite. (Read more about [Redwood Contributors here](https://community.redwoodjs.com/t/contributors-meetups-a-next-step-for-redwoodjs-participation/2470).)

Could you send me a DM on the Redwood Forums? Or any other preferred way to connect?

https://community.redwoodjs.com/u/thedavid/

**Pull request merged**